ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -879,13 +999,109 @@ applyByronGenTx validationMode (ByronEBBLedgerConfig (ByronLedgerConfig cfg)) ge
+          , V.Interface.k                 = k
+          , V.Interface.currentEpoch      = currentEpoch
+          , V.Interface.currentSlot       = currentSlot
+          }
+
+        fixPM (Crypto.AProtocolMagic a b) =
+          Crypto.AProtocolMagic (reAnnotate a) b
+
+    -- Note that we're only expecting update-related 'GenTx's in this case.
+    -- Both of them share much of the same code so it didn't seem to make
+    -- sense defining two separate cases of 'go'.
+    go tx cvs = wrapCVS <$> case tx of
+        ByronUpdateProposal _ proposal ->
+          CC.UPI.registerProposal env updateState proposal
+            `wrapError` ByronApplyUpdateProposalError
+
+        ByronUpdateVote _ vote ->
+          CC.UPI.registerVote env updateState vote
+            `wrapError` ByronApplyUpdateVoteError
+
+        _ -> error $  "applyByronGenTx.go: "
+                   <> "Only expecting update-related `GenTx`s at this point."
+      where
+        wrapCVS updateState' = cvs { CC.Block.cvsUpdateState = updateState' }
+
+        protocolMagic = Crypto.getAProtocolMagicId $ fixPM $
+          CC.Genesis.configProtocolMagic cfg
+
+        k = CC.Genesis.configK cfg
+
+        updateState = CC.Block.cvsUpdateState cvs
+
+        currentSlot = CC.Block.cvsLastSlot cvs
+
+        numGenKeys = toNumGenKeys $ Set.size (allowedDelegators cfg)
+
+        delegationMap =
+          (V.Interface.delegationMap . CC.Block.cvsDelegationState) cvs
+
+        env = CC.UPI.Environment
+          { CC.UPI.protocolMagic = protocolMagic
+          , CC.UPI.k             = k
+          , CC.UPI.currentSlot   = currentSlot
+          , CC.UPI.numGenKeys    = numGenKeys
+          , CC.UPI.delegationMap = delegationMap
+          }
+
+        fixPM (Crypto.AProtocolMagic a b) =
+          Crypto.AProtocolMagic (reAnnotate a) b
+
+        toNumGenKeys :: Integral n => n -> Word8
+        toNumGenKeys n
+          | n > fromIntegral (maxBound :: Word8) = error $
+            "applyByronGenTx.go.toNumGenKeys: Too many genesis keys"
+          | otherwise = fromIntegral n
+
+mkByronGenTx :: CC.Mempool.AMempoolPayload ByteString
+             -> GenTx (ByronBlockOrEBB cfg)
+mkByronGenTx mp = case mp of
+    CC.Mempool.MempoolTx tx ->
+      ByronTx (Crypto.hash $ CC.UTxO.taTx tx)  -- TODO replace this with a
+                                               -- function from cardano-ledger,
+                                               -- see cardano-ledger#581
+              tx
+
+    CC.Mempool.MempoolDlg cert ->
+      ByronDlg (Crypto.hash $ void cert) cert
+
+    CC.Mempool.MempoolUpdateProposal proposal ->
+      ByronUpdateProposal (CC.Update.recoverUpId proposal) proposal
+
+    CC.Mempool.MempoolUpdateVote vote ->
+      ByronUpdateVote (Crypto.hash $ void vote) vote
+
+mkMempoolPayload :: GenTx (ByronBlockOrEBB cfg)
+                 -> CC.Mempool.AMempoolPayload ByteString
+mkMempoolPayload genTx = case genTx of
```

Comment:
Could we use `CC.Mempool.AMempoolPayload ByteString` instead of redefining it in this module? I guess it's just the cached `GenTxId`s that are preventing us from doing this?

Reply:
Yes, it's as you said. Since we're maintaining the cached `GenTxId`s we can't do this. Even if I add a `MempoolPayloadId` and define something like:

```
data GenTx (ByronBlockOrEBB cfg) = ByronGenTx (AMempoolPayloadId ByteString) !(AMempoolPayload ByteString)
```

it has the potential to get messy as you could have  `MempoolPayloadId`s and `MempoolPayload`s which don't correctly correspond to each other. e.g. you could have a value like `ByronGenTx MempoolDlgId (MempoolUpdateVote vote)`.

Reply:
I could define something like `MempoolPayloadAndId` (basically the same as this existing definition for `GenTx`) in `cardano-ledger` but I feel like the cached `GenTxId`s are a bit of a implementation detail in here `ouroboros-consensus` rather than something that ledger should care about.